contrib/cirrus/runner.sh

```diff
@@ -345,6 +364,51 @@ dotest() {
         |& logformatter
 }
 
+# Optimization: will exit if the only PR diffs are under docs/ or tests/
+# with the exception of any given arguments. E.g., don't run e2e or upgrade
+# or bud tests if the only PR changes are in test/system.
+function _bail_if_test_can_be_skipped() {
+    local head base diffs
+
+    # Cirrus sets these for PRs but not cron. In cron, we never want to skip.
+    for v in CIRRUS_CHANGE_IN_REPO DEST_BRANCH; do
+        if [[ -z "${!v}" ]]; then
+            echo "[ _cannot do selective skip: \$$v is undefined ]"
+            return 0
+        fi
+    done
+    # And if this one *is* defined, it means we're not in PR-land; don't skip.
+    if [[ -n "$CIRRUS_TAG" ]]; then
+        echo "[ _cannot do selective skip: \$CIRRUS_TAG is defined ]"
+        return 0
+    fi
+
+    head=$CIRRUS_CHANGE_IN_REPO
+    base=$(git merge-base $DEST_BRANCH $head)
+    diffs=$(git diff --name-only $base $head)
+
+    # If PR touches any files in an argument directory, we cannot skip
+    for subdir in "$@"; do
+        if egrep -q "^$subdir/" <<<"$diffs"; then
+            return 0
+        fi
+    done
+
+    # Filter these out from the diffs, see what's left.
```

Comment:
This comment (WRT `for subdir in docs test; do`) is confusing to me (it specifies "what" not "why").  Shouldn't it be filtering out based on the arguments (`$@`) not hard-coded values?

Reply:
Arguments have already been filtered out (in the code immediately above). What we do now is: if this is a docs-or-test-only PR, then it's safe to skip. The `docs` and `test` hardcoding has to happen here, not in the caller: specifying them in the caller was my approach last night, and it was a nightmare.

Reply:
So if you remember, for a future PR or update, please fix the comment to mention something like "hard-coding docs and test to avoid overburdening caller".  Or maybe in the main function comment.